File: algorithms/459.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
private:
    vector<int> kmp(const string &s) {
        int n = s.size();
        vector<int> lps(n, 0);
        for (int i = 1, len = 0; i < n; ) {
            if (s[i] == s[len])
                lps[i++] = ++len;
            else if (len)
                len = lps[len-1];
            else
                lps[i++] = 0;
        }
        return lps;
    }
public:
    bool repeatedSubstringPattern(string s) {
        int n = s.size();
        vector<int> lps = kmp(s);
        int len = lps[n-1];
        return len > 0 && (n % (n-len) == 0);
    }
};

int main() {
    return 0;
}

File: algorithms/rust/459.rs
impl Solution {
    #[allow(dead_code)]
    pub fn repeated_substring_pattern(s: String) -> bool {
        let n = s.len();

        let s_bytes = s.as_bytes();
        for i in 1..=(n / 2) {
            if n % i != 0 {
                continue;
            }
            let mut flag = true;
            for (idx, &ch) in s_bytes.iter().enumerate() {
                if ch != s_bytes[idx % i] {
                    flag = false;
                    break;
                }
            }
            if flag {
                return true;
            }
        }
        false
    }
}

struct Solution;

fn main() {
    println!("Hello, world!");
}
